server/protocols/jwt/src/main/java/org/apache/james/jwt/JwtTokenVerifier.java

```diff
@@ -67,12 +99,19 @@ public static JwtTokenVerifier create(JwtConfiguration jwtConfiguration) {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(JwtTokenVerifier.class);
 
+    private final JwtParser kidJwtParser;
     private final List<JwtParser> jwtParsers;
 
     public JwtTokenVerifier(PublicKeyProvider pubKeyProvider) {
+        // one parser that performs key lookup by kid
+        this.kidJwtParser = toImmutableJwtParser(jwtHeaders -> {
+            String kid = Objects.requireNonNull((String)jwtHeaders.get("kid"));
+            return pubKeyProvider.get(kid).orElse(null);
```

Comment:
Before we were loading public keys only at startup.

Now we are loading keys on each and every token we verify. Thus I would not be surprised performance significantly degrades.

Maybe we could load upfront a `Map<Kid -> PublicKey> in the DefaultPublicKeyProvider in order to keep this desirable property?

Maybe the

Reply:
Perhaps I'm missing something, but aiui OidcJwtTokenVerifier (the only place currently using kid) creates a new provider instance per request, so any caching within the providers would be meaningless? It seems like the providers themselves need to be cached instead of using stateless static methods, but that goes back to other components I haven't gone over here. Maybe I'm assuming something wrong about the other components using this verifier, or they are not called per request etc.?